cast between incompatible function types

gcc8 on macos:

```
serv.c:29:20: error: cast between incompatible function types from 'void (*)(Conn *, int)' {aka 'void (*)(struct Conn *, int)'} to 'void (*)(void *, size_t)' {aka 'void (*)(void *, long unsigned int)'} [-Werror=cast-function-type]
     s->conns.rec = (Record)connrec;
                    ^
```

The type of `Serv.conns` is Heap while `connrec` sets `tickpos` to -1 for some reason. Needs further investigation.

--- a/conn.c
+++ b/conn.c
@@ -52,7 +52,8 @@ make_conn(int fd, char start_state, tube use, tube watch)
     c->sock.fd = fd;
     c->state = start_state;
     c->pending_timeout = -1;
-    c->tickpos = -1;
+    c->tickpos = 0; // Does not mean anything if in_conns is set to 0.
+    c->in_conns = 0;
     j = &c->reserved_jobs;
     j->prev = j->next = j;
 
@@ -110,6 +111,7 @@ has_reserved_job(Conn *c)
 }
 
 
+// Returns positive nanoseconds when c should tick, 0 otherwise.
 static int64
 conntickat(Conn *c)
 {
@@ -136,6 +138,8 @@ conntickat(Conn *c)
 }
 
 
+// TODO: remove this function by inlining its content into 3 callees places.
+// Reason: conn.c does not use rw anywhere in this file.
 void
 connwant(Conn *c, int rw)
 {
@@ -144,21 +148,25 @@ connwant(Conn *c, int rw)
 }
 
 
+// Remove c from the c->srv heap and reschedule it using the value
+// returned by conntickat if there is an outstanding timeout in the c.
 void
 connsched(Conn *c)
 {
-    if (c->tickpos > -1) {
+    if (c->in_conns) {
         heapremove(&c->srv->conns, c->tickpos);
+        c->in_conns = 0;
     }
     c->tickat = conntickat(c);
     if (c->tickat) {
         heapinsert(&c->srv->conns, c);
+        c->in_conns = 1;
     }
 }
 
 
-/* return the reserved job with the earliest deadline,
- * or NULL if there's no reserved job */
+// Return the reserved job with the earliest deadline,
+// or NULL if there's no reserved job
 job
 connsoonestjob(Conn *c)
 {
@@ -167,7 +175,8 @@ connsoonestjob(Conn *c)
 
     if (soonest == NULL) {
         for (j = c->reserved_jobs.next; j != &c->reserved_jobs; j = j->next) {
-            if (j->r.deadline_at <= (soonest ? : j)->r.deadline_at) soonest = j;
+            if (j->r.deadline_at <= (soonest ? : j)->r.deadline_at)
+                soonest = j;
         }
     }
     c->soonest_job = soonest;
@@ -175,8 +184,8 @@ connsoonestjob(Conn *c)
 }
 
 
-/* return true if c has a reserved job with less than one second until its
- * deadline */
+// Return true if c has a reserved job with less than one second until its
+// deadline.
 int
 conndeadlinesoon(Conn *c)
 {
@@ -192,7 +201,8 @@ conn_ready(Conn *c)
     size_t i;
 
     for (i = 0; i < c->watch.used; i++) {
-        if (((tube) c->watch.items[i])->ready.len) return 1;
+        if (((tube) c->watch.items[i])->ready.len)
+            return 1;
     }
     return 0;
 }
@@ -206,7 +216,7 @@ connless(Conn *a, Conn *b)
 
 
 void
-connrec(Conn *c, int i)
+connrec(Conn *c, size_t i)
 {
     c->tickpos = i;
 }
@@ -241,8 +251,9 @@ connclose(Conn *c)
     c->use->using_ct--;
     TUBE_ASSIGN(c->use, NULL);
 
-    if (c->tickpos > -1) {
+    if (c->in_conns) {
         heapremove(&c->srv->conns, c->tickpos);
+        c->in_conns = 0;
     }
 
     free(c);

--- a/dat.h
+++ b/dat.h
@@ -275,9 +275,10 @@ struct Conn {
     char   state;
     char   type;
     Conn   *next;
-    tube   use;
-    int64  tickat;      // time at which to do more work
-    int    tickpos;     // position in srv->conns
+    tube   use;         // tube currently in use
+    int64  tickat;      // time at which to do more work; determines pos in heap
+    size_t tickpos;     // position in srv->conns, stale when in_conns=0
+    byte   in_conns;    // 1 if the conn is in srv->conns heap, 0 otherwise
     job    soonest_job; // memoization of the soonest job
     int    rw;          // currently want: 'r', 'w', or 'h'
     int    pending_timeout;
@@ -306,7 +307,7 @@ struct Conn {
     struct job reserved_jobs; // linked list header
 };
 int  connless(Conn *a, Conn *b);
-void connrec(Conn *c, int i);
+void connrec(Conn *c, size_t i);
 void connwant(Conn *c, int rw);
 void connsched(Conn *c);
 void connclose(Conn *c);
@@ -387,6 +388,8 @@ struct Server {
 
     Wal    wal;
     Socket sock;
+
+    // Connections that must produce deadline or timeout, ordered by the time.
     Heap   conns;
 };
 void srvserve(Server *srv);

--- a/prot.c
+++ b/prot.c
@@ -1864,6 +1864,7 @@ prothandle(Conn *c, int ev)
     h_conn(c->sock.fd, ev, c);
 }
 
+// prottick returns nanoseconds till the next work.
 int64
 prottick(Server *s)
 {
@@ -1899,6 +1900,8 @@ prottick(Server *s)
         }
     }
 
+    // Process connections with pending timeouts. Release jobs with expired ttr.
+    // Capture the period from the soonest connection.
     while (s->conns.len) {
         Conn *c = s->conns.data[0];
         d = c->tickat - now;
@@ -1908,6 +1911,7 @@ prottick(Server *s)
         }
 
         heapremove(&s->conns, 0);
+        c->in_conns = 0;
         conn_timeout(c);
     }
 

